File: app/motion_detector.py
import cv2
import numpy as np
from typing import Final, Optional

from app.point import Point


class MotionDetector:

    THRESHOLD_VALUE: Final[int] = 30
    MIN_CONTOUR_AREA: Final[int] = 200
    DILATE_ITERATIONS: Final[int] = 2

    def __init__(self, mask: Optional[np.ndarray] = None):
        self.mask = mask

    def detect(self, prev_img: np.ndarray, curr_img: np.ndarray) -> list[np.ndarray]:
        diff = cv2.absdiff(prev_img, curr_img)
        gray = cv2.cvtColor(diff, cv2.COLOR_BGR2GRAY)

        _, thresh = cv2.threshold(gray, self.THRESHOLD_VALUE, 255, cv2.THRESH_BINARY)
        dilated = cv2.dilate(thresh, None, iterations=self.DILATE_ITERATIONS)

        if self.mask is not None:
            dilated = cv2.bitwise_and(dilated, self.mask)

        contours, _ = cv2.findContours(
            dilated, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE
        )

        if not contours:
            return []

        filtered = [c for c in contours if cv2.contourArea(c) >= self.MIN_CONTOUR_AREA]

        if not filtered:
            return []

        filtered.sort(key=lambda c: cv2.contourArea(c), reverse=True)
        return filtered

    def debug(self, contours, image):
        output = image.copy()

        for c in contours:
            area = cv2.contourArea(c)
            cv2.drawContours(output, [c], -1, (0, 255, 0), 2)

            x, y, w, h = cv2.boundingRect(c)
            text_pos = (x, y - 5 if y - 5 > 0 else y + 15)

            cv2.putText(
                output,
                f"{int(area)}",
                text_pos,
                cv2.FONT_HERSHEY_SIMPLEX,
                0.6,
                (0, 0, 255),
                2,
                cv2.LINE_AA,
            )

        cv2.imwrite("output.png", output)

File: App/MotionDetector.cs
using DekapuAutoOpencv.Models;
using OpenCvSharp;

using OcvPoint = OpenCvSharp.Point;

namespace DekapuAutoOpencv.Core;

public class MotionDetector(AppSettings settings, Mat? mask = null)
{
    public OcvPoint[][] Detect(Mat prevImg, Mat currImg)
    {
        using var diff = new Mat();
        using var gray = new Mat();
        using var thresh = new Mat();
        using var dilated = new Mat();

        Cv2.Absdiff(prevImg, currImg, diff);
        Cv2.CvtColor(diff, gray, ColorConversionCodes.BGR2GRAY);
        Cv2.Threshold(gray, thresh, settings.ThresholdValue, 255, ThresholdTypes.Binary);
        Cv2.Dilate(thresh, dilated, null, iterations: settings.DilateIterations);

        if (mask != null)
        {
            if (mask.Size() != dilated.Size())
                Cv2.Resize(mask, mask, dilated.Size(), 0, 0, InterpolationFlags.Nearest);

            Cv2.BitwiseAnd(dilated, mask, dilated);
        }

        Cv2.FindContours(
            dilated,
            out var contours,
            out _,
            RetrievalModes.External,
            ContourApproximationModes.ApproxSimple);

        if (contours.Length == 0)
            return [];

        var filtered = contours
            .Where(c => Cv2.ContourArea(c) >= settings.MinContourArea)
            .OrderByDescending(c => Cv2.ContourArea(c))
            .ToArray();

        return filtered;
    }

    public static Mat Annotate(OcvPoint[][] contours, Mat image)
    {
        var output = image.Clone();

        foreach (var c in contours)
        {
            var rect = Cv2.BoundingRect(c);
            var area = (int)Cv2.ContourArea(c);

            Cv2.DrawContours(output, [c], -1, new Scalar(0, 255, 0), 2);

            const double fontSize = 0.55;
            const int fontThickness = 1;
            var label = $"{area}px ({rect.Width}x{rect.Height})";
            var textSize = Cv2.GetTextSize(label, HersheyFonts.HersheySimplex, fontSize, fontThickness, out var baseline);

            var textOrg = new OcvPoint(rect.X, rect.Y > textSize.Height + 4 ? rect.Y - 4 : rect.Y + textSize.Height + 4);
            var bgTl = new OcvPoint(textOrg.X - 1, textOrg.Y - textSize.Height - 1);
            var bgBr = new OcvPoint(textOrg.X + textSize.Width + 1, textOrg.Y + baseline + 1);

            Cv2.Rectangle(output, bgTl, bgBr, new Scalar(255, 255, 255), -1);
            Cv2.PutText(output, label, textOrg, HersheyFonts.HersheySimplex, fontSize, new Scalar(0, 0, 0), fontThickness, LineTypes.AntiAlias);
        }

        return output;
    }
}
